doc/builders/builders.chapter.md

```diff
@@ -0,0 +1,42 @@
+# The builders {#chap-builders}
+
+A Builder, in the context of Nixpkgs, are functions that produces derivations. (Don't confuse it with the `builder` input argument of function `derivation`, which refers to the path to the executable that builds the derivation.) Such function is usually designed to follow a specific workflow, so that one can produce a package by specifying a few options instead of messing up with the `derivation` function.
```

Comment:
Possible candidates:

*   Build function
    *   Existing use referring to the same thing
        *   `androidenv.buildApp` `doc/languages-frameworks/android.section.md` `## Building an Android application {#building-an-android-application}`
            > This build function is particularly useful when it is desired to use
[Hydra](https://nixos.org/hydra): the Nix-based continuous integration solution
to build Android apps.

        *   `xcodeenv.composeXcodeWrapper` `doc/languages-frameworks/ios.section.md` `## Deploying a proxy component wrapper exposing Xcode {#deploying-a-proxy-component-wrapper-exposing-xcode}`
            > The first use case is deploying a Nix package that provides symlinks to the Xcode
installation on the host system. This package can be used as a build input to
any build function implemented in the Nix expression language that requires
Xcode.

    *   Existing use referring to something else
        *   Inside the build process of `pkgs/build-support/rust/build-rust-crate/build-crate.nix`
            > \# configure & source common build functions

*   Build helper
    *   Existing use referring to the same thing
        *   `buildIntegration` `pkgs/tools/networking/dd-agent/integrations-core.nix`
            > \# Build helper to build a single datadog integration package.

    *   Existing use referring to something else
        *   Inside the `meta.description` of the Haskell Hackage package `hein` `pkgs/development/haskell-modules/hackage-packages.nix`
            > An extensible build helper for haskell, in the vein of leiningen

The different use of both candidates doesn't seem to cause confusion. We could just choose from one of them, or come up with something else.

Reply:
Great research. I like "build helper" more, because its closest to what the thing is about, and also "build function" only contains one word that adds information (almost everything is a function).

When we change section titles we should make sure URLs stay stable. Either keep the anchors or add redirects (we discussed this with @pennae somewhat recently but I can't remember why the script used in the Nix manual was not included).

Reply:
I disagree that we should rename this from builder to something else and that "build helper" is better. Quite the contrary, the function when called actually builds something, which I'd say is exactly what a builder is. This naming has been around for a *very* long time.

The Nix attribute `builder` is a detail that no typical user would *ever* encounter. Those interested going through say the Pills or in detail through the Nix manual will see it, but otherwise unless you're interested in writing a custom `stdenv.mkDerivation` you won't do anything with it. Hence, I don't think that part should even be mentioned here.

Reply:
Following that line of reasoning "build helper" is also incorrect, because we're solely preprocessing before deriving. So in that case "deriver" or "deriver helper" or "build planner" is more correct.

Reply:
Clearly this needs discussion, but I don't want to block this PR on a single term. I'm happy with "build helper" because it reduces ambiguity and is not obviously wrong, and I'm not opposed to further improving the terminology in a separate PR.

Reply:
Maybe we could move the `builder` naming to another issue or PR.

BTW, "builders" also refers to "build machines" in the context of "remote builders", which is probably why the documentation of `darwin.builder` is misplaced under the "Builders" part in the Nixpkgs manual. It would be great to find a more suitable place for it.

Reply:
Did we ever have a user be confused about these being called builders? I at least haven't, and I don't think we should be even spending time discussing this unless there are users confused about it.